src/ILCompiler.Compiler/src/Compiler/TypePreinit.cs

```diff
@@ -1698,21 +1836,86 @@ public override void WriteFieldData(ref ObjectDataBuilder builder, FieldDesc fie
             }
         }
 
-        private abstract class ReferenceTypeValue : Value, ISerializableReference
+        private abstract class ReferenceTypeValue : Value
         {
-            protected readonly TypeDesc _type;
+            public TypeDesc Type { get; }
 
-            protected ReferenceTypeValue(TypeDesc type) { _type = type; }
+            protected ReferenceTypeValue(TypeDesc type) { Type = type; }
 
             public override bool Equals(Value value)
             {
                 return this == value;
             }
+        }
+
+        private class DelegateInstance : ReferenceTypeValue, ISerializableReference
+        {
+            private readonly MethodDesc _methodPointed;
+            private readonly ForeignTypeInstance _firstParameter;
+
+            public DelegateInstance(TypeDesc delegateType, MethodDesc methodPointed, ForeignTypeInstance firstParameter)
+                : base(delegateType)
+            {
+                _methodPointed = methodPointed;
+                _firstParameter = firstParameter;
+            }
+
+            public virtual void WriteContent(ref ObjectDataBuilder builder, ISymbolNode thisNode, NodeFactory factory)
+            {
+                Debug.Assert(_methodPointed.Signature.IsStatic == (_firstParameter == null));
+                Debug.Assert(!_methodPointed.IsUnmanagedCallersOnly);
+
+                var creationInfo = DelegateCreationInfo.Create(Type.ConvertToCanonForm(CanonicalFormKind.Specific), _methodPointed, factory, followVirtualDispatch: false);
 
-            public abstract void WriteContent(ref ObjectDataBuilder builder, NodeFactory factory);
+                Debug.Assert(!creationInfo.TargetNeedsVTableLookup);
+
+                // EEType
+                var node = factory.ConstructedTypeSymbol(Type);
+                Debug.Assert(!node.RepresentsIndirectionCell);  // Shouldn't have allowed this
+                builder.EmitPointerReloc(node);
+
+                if (_methodPointed.Signature.IsStatic)
+                {
+                    Debug.Assert(creationInfo.Constructor.Method.Name == "InitializeOpenStaticThunk");
```

Comment:
I have noticed that quite a few of Initialize methods on Delegate are unused:
```
InitializeClosedStaticWithoutThunk
InitializeOpenStaticWithoutThunk
InitializeReversePInvokeWithoutThunk
InitializeOpenInstanceThunk
InitializeOpenInstanceWithoutThunk
```
You can delete them while you are on it.